### PR TITLE
Refactor Sync API

### PR DESCRIPTION
**Pull request by jingtang10**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

**Description**
1. Rename `FhirPeriodicSyncWorker` to `FhirSyncWorker` since it is now used by both periodic sync and one time sync
2. Remove the `SyncJob` API and its implementation `SyncJobImp` since it is not the right level of abstraction we should be providing to the application.
3. Use `StateFlow` to return the worker status for both one time and periodic work.
4. Use the worker name to uniquely identify a job to allow multiple jobs to be scheduled. 
5. Stop leaking the internal result to external APIs and rename `State` to `SyncJobStatus`

Future work:
1. Still need to emit the last sync timestamp for the application to listen to.

**Alternative(s) considered**
@aditya-07 and I have discussed the current approach and concluded the current API isn't easy to use by the client

**Type**
Code health

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

**Comment by jingtang10** [comment]
@maimoonak FYI - the current sync APIs need some simplification. please see this pr. i think the main issue is that the sync job shouldn't be exposed to the client and information should be exposed as flow. we can discuss in more detail if you like.

@joiskash and @PallaviGanorkar fyi too

discussed this with @aditya-07

File: engine/src/main/java/com/google/android/fhir/sync/Sync.kt [hunk 1]
```diff
@@ -75,22 +56,40 @@ object Sync {
...
+            ?.let {
+              val state = it.getString("StateType")!!
+              val stateData = it.getString("State")
+              gson.fromJson(stateData, Class.forName(state)) as State
+            }
+        }
 
     WorkManager.getInstance(context)
       .enqueueUniquePeriodicWork(
         SyncWorkType.DOWNLOAD.workerName,
```
[the remark this answers is not in the document]
Comment (jingtang10) [reply engine/src/main/java/com/google/android/fhir/sync/Sync.kt]: done

File: engine/src/main/java/com/google/android/fhir/sync/Sync.kt [hunk 2]
```diff
@@ -82,32 +48,52 @@ object Sync {
...
-  ) {
-
+  ): Flow<State> {
+    val flow = getStateFlow<W>(context)
     WorkManager.getInstance(context)
       .enqueueUniquePeriodicWork(
-        SyncWorkType.DOWNLOAD.workerName,
-        ExistingPeriodicWorkPolicy.KEEP,
+        W::class.java.name,
+        ExistingPeriodicWorkPolicy.REPLACE,
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/sync/Sync.kt:76]: `ExistingPeriodicWorkPolicy.REPLACE` may cause a running task to get canceled. see https://developer.android.com/topic/libraries/architecture/workmanager/how-to/managing-work#stop-worker .

File: engine/src/main/java/com/google/android/fhir/sync/Sync.kt [hunk 3]
```diff
@@ -82,32 +48,52 @@ object Sync {
...
         createPeriodicWorkRequest(periodicSyncConfiguration, W::class.java)
       )
+    return flow
   }
 
+  inline fun <reified W : FhirSyncWorker> getStateFlow(context: Context) =
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/sync/Sync.kt:82]: We should encourage users to use this to collect state of an already queued/running work. Maybe add some docs to this.

Comment (jingtang10) [reply engine/src/main/java/com/google/android/fhir/sync/Sync.kt:82]: ok i renamed this api and adde documentation

File: engine/src/main/java/com/google/android/fhir/sync/FhirSynchronizer.kt [hunk 4]
```diff
@@ -72,6 +72,8 @@ internal class FhirSynchronizer(
   }
 
   private suspend fun setSyncState(result: Result): Result {
+
+    // todo: emit this properly instead of using datastore?
     datastoreUtil.writeLastSyncTimestamp(result.timestamp)
```
[the remark this answers is not in the document]
Comment (jingtang10) [reply engine/src/main/java/com/google/android/fhir/sync/FhirSynchronizer.kt:66]: done

File: demo/src/main/java/com/google/android/fhir/demo/PatientListFragment.kt [hunk 5]
```diff
@@ -153,24 +147,24 @@ class PatientListFragment : Fragment() {
...
             Timber.i("Sync: ${it::class.java.simpleName} with ${it.resourceType?.name}")
             fadeInTopBanner()
           }
-          is State.Finished -> {
-            Timber.i("Sync: ${it::class.java.simpleName} at ${it.result.timestamp}")
+          is SyncJobStatus.Finished -> {
+            Timber.i("Sync: ${it::class.java.simpleName} at ${it.timestamp}")
             patientListViewModel.searchPatientsByName(searchView.query.toString().trim())
-            mainActivityViewModel.updateLastSyncTimestamp()
+            // mainActivityViewModel.updateLastSyncTimestamp()
```
Comment (aditya-07) [inline demo/src/main/java/com/google/android/fhir/demo/PatientListFragment.kt:161]: Should this be uncommented?

File: demo/src/main/java/com/google/android/fhir/demo/PatientListFragment.kt [hunk 6]
```diff
@@ -153,24 +147,24 @@ class PatientListFragment : Fragment() {
...
             fadeOutTopBanner(it)
           }
-          is State.Failed -> {
-            Timber.i("Sync: ${it::class.java.simpleName} at ${it.result.timestamp}")
+          is SyncJobStatus.Failed -> {
+            Timber.i("Sync: ${it::class.java.simpleName} at ${it.timestamp}")
             patientListViewModel.searchPatientsByName(searchView.query.toString().trim())
-            mainActivityViewModel.updateLastSyncTimestamp()
+            // mainActivityViewModel.updateLastSyncTimestamp()
```
Comment (aditya-07) [inline demo/src/main/java/com/google/android/fhir/demo/PatientListFragment.kt:167]: same

File: engine/src/main/java/com/google/android/fhir/sync/SyncJobStatus.kt [hunk 7]
```diff
@@ -16,18 +16,15 @@
...
-  override fun getFhirEngine() = TestingUtils.TestFhirEngineImpl
-
-  override fun getDownloadWorkManager() = TestingUtils.TestDownloadManagerImpl()
-
-  override fun getConflictResolver() = AcceptRemoteConflictResolver
+  class Started : SyncJobStatus()
+  data class InProgress(val resourceType: ResourceType?) : SyncJobStatus()
+  data class Glitch(val exceptions: List<ResourceSyncException>) : SyncJobStatus()
+  class Finished : SyncJobStatus()
+  data class Failed(val exceptions: List<ResourceSyncException>) : SyncJobStatus()
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/sync/SyncJobStatus.kt:29]: We should add some docs to explain the various statuses.

Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/sync/SyncJobStatus.kt:29]: Also, add a single line in between the classes.

File: engine/src/main/java/com/google/android/fhir/sync/Sync.kt (hunk 2 above)
Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/sync/Sync.kt:87]: I think so.

Comment (aditya-07) [reply engine/src/main/java/com/google/android/fhir/sync/Sync.kt:87]: Should we provide an optional param to the users to override the behaviour of keeping or replacing the current running one?

File: demo/src/main/java/com/google/android/fhir/demo/PatientListFragment.kt (hunk 5 above)
Comment (jingtang10) [reply demo/src/main/java/com/google/android/fhir/demo/PatientListFragment.kt:161]: oops. thanks for catching.

File: demo/src/main/java/com/google/android/fhir/demo/PatientListFragment.kt (hunk 6 above)
Comment (jingtang10) [reply demo/src/main/java/com/google/android/fhir/demo/PatientListFragment.kt:167]: ditto

File: engine/src/main/java/com/google/android/fhir/sync/SyncJobStatus.kt (hunk 7 above)
Comment (jingtang10) [reply engine/src/main/java/com/google/android/fhir/sync/SyncJobStatus.kt:29]: done

File: engine/src/main/java/com/google/android/fhir/sync/Sync.kt (hunk 2 above)
Comment (jingtang10) [reply engine/src/main/java/com/google/android/fhir/sync/Sync.kt]: done. i think we can introduce the param in a follow up

File: demo/src/main/java/com/google/android/fhir/demo/MainActivity.kt [hunk 8]
```diff
@@ -78,7 +80,7 @@ class MainActivity : AppCompatActivity() {
   private fun onNavigationItemSelected(item: MenuItem): Boolean {
     when (item.itemId) {
       R.id.menu_sync -> {
-        viewModel.poll()
+        Sync.oneTimeSync<FhirSyncWorker>(this)
```
[the remark this answers is not in the document]
Comment (jingtang10) [reply demo/src/main/java/com/google/android/fhir/demo/MainActivity.kt]: done

**Review by aditya-07 — APPROVED** [review]
LGTM!

**Comment by jingtang10** [comment]
#1307

**Pull request merged**